src/wp-includes/fonts.php

```diff
@@ -131,7 +131,44 @@ function wp_get_font_dir() {
 	 *
 	 * @param array $defaults The original fonts directory data.
 	 */
-	return apply_filters( 'font_dir', $defaults );
+	$font_dir = apply_filters( 'font_dir', $default_dir );
+
+	wp_mkdir_p( $font_dir['path'] );
```

Comment:
This function can not attempt to create the directory. It's unaware if `wp_upload_dir()` has been called to get the directory without creating it.

If the directory doesn't exist, we'll need to check if the parent directory is writable.

Reply:
As far as I see `wp_mkdir_p()` expects full paths and makes directories recursively (calls PHP's `mkdir()` with `true` as third param). That should make any missing parent and sub-parent directories? I.e. trying to make `/wp-content/dir1/dir2/fonts` would work, and `dir1` and `dir2` will be created if they don't exist.

> If the directory doesn't exist, we'll need to check if the parent directory is writable.

May be missing something but seems `wp_mkdir_p()` would return `false` if the parent (or grandparent, great-grandparent, etc.) directory is not writable. Seems that's how it is used elsewhere in core? Then checking beforehand whether the parent directory is writable sounds like a tiny bit of optimization for the cases where that directory is not writable, but a tiny bit of overhead for cases where it is writable (more common)? 

I.e. something like this code will have to run before every `wp_mkdir_p()` call:
```
// Do not allow path traversals.
if ( str_contains( $target, '../' ) || str_contains( $target, '..' . DIRECTORY_SEPARATOR ) ) {
	return false; // Bail out somehow...
}

// We need to find the permissions of the parent folder that exists and inherit that.
$target_parent = dirname( $target );
while ( '.' !== $target_parent && ! is_dir( $target_parent ) && dirname( $target_parent ) !== $target_parent ) {
	$target_parent = dirname( $target_parent );
}

// Get the permission bits.
$stat = @stat( $target_parent );
// Or maybe use is_writable( $target_parent ); (seems this might be a bit problematic on some filesystems if $target_parent was just created)
```

And then almost the same code will run inside `wp_mkdir_p()` if the permissions indicate that the directory is writable.

Frankly I'm not sure that the extra code to check the parent directory's permissions is warranted?

Reply:
> Frankly I'm not sure that the extra code to check the parent directory's permissions is warranted? Seems `wp_mkdir_p()` can handle the case where parent is not writable?

`wp_upload_dir()` (which this code filters) runs in two context: one in which it creates the directory, one in which it doesn't.

As this function doesn't know the context re file creation, it's not able to create a directory as it risks adding junk directories to the server. Junk meaning that they may never be used.

The font directory functions are also used as a getter during post deletion so we can't create directories to delete files that do not exist.

Reply:
Sorry but I think I'm really missing something here.

> wp_upload_dir() (which this code filters) runs in two context: one in which it creates the directory, one in which it doesn't.

Right.

> As this function doesn't know the context re file creation

I.e. we don't know whether `wp-content` or `wp-content/uploads` (or their replacements if somebody moved/renamed them) exist, correct? But how could that be after `wp_upload_dir()` has already returned the path and URL to the uploads dir?

> it's not able to create a directory as it risks adding junk directories to the server

Meaning `wp_mkdir_p()` (actually PHP's `mkdir()`) will not be able to create a directory but may add "junk" directories? Sorry but I'm lost, how can it add directories without being able to create them...

Perhaps some practical example would help? Also, what would a fix for this look like? Maybe that would help to understand and recreate the problem?

Reply:
It has nothing to do with whether `wp_mkdir_p()` can create the directory.

What I want to avoid is this function attempting to create the directories when it should not. This can happen in two circumstances:

* when hooked on to `upload_dir`: [`wp_upload_dir()`](https://developer.wordpress.org/reference/functions/wp_upload_dir/) is called with the `$create_dir` parameter set to false, either directly or via [`wp_get_upload_dir()`](https://developer.wordpress.org/reference/functions/wp_get_upload_dir/).
* when called as a getter in [`_wp_before_delete_font_face()`](https://github.com/WordPress/wordpress-develop/blob/9c415ed5fb9b4069208b9e43ff774acbfb3ef3a6/src/wp-includes/fonts.php#L163-L183): this is used to delete font files if the `_wp_font_face_file` post meta is defined. As font face post types may not contain a font file (see the discussion in https://github.com/WordPress/gutenberg/issues/59387) then WordPress can't start creating directories in order to delete files that don't exist.

Reply:
Ah I see what you mean. So basically this function has to be split in two: a getter and then a function that would only be used to create the dirs if needed.

Reply:
@azaozz Yes, that's correct. 

It's what I did in the PR to fix the infinite loop issue rather than work aroudn it https://github.com/WordPress/wordpress-develop/pull/6211 (reviews welcome, btw), partly to prepare for this.